Stratis.Bitcoin/Features/BlockStore/BlockStoreLoop.cs

```diff
@@ -1,347 +1,187 @@
+        ///     2: Download missing blocks and write to store
+        /// </summary>
+		internal void StartLoop()
+        {
+            this.asyncLoopFactory.Run("BlockStoreLoop.DownloadBlocks", async token =>
+                {
+                    await DownloadAndStoreBlocks(this.nodeLifetime.ApplicationStopping);
+                },
+                this.nodeLifetime.ApplicationStopping,
+                repeatEvery: TimeSpans.Second,
+                startAfter: TimeSpans.FiveSeconds);
+        }
 
-		public async Task DownloadAndStoreBlocks(CancellationToken token, bool disposemode = false)
-		{
+        private async Task DownloadAndStoreBlocks(CancellationToken cancellationToken, bool disposeMode = false)
+        {
             // TODO: add support to BlockStoreLoop to unset LazyLoadingOn when not in IBD
             // When in IBD we may need many reads for the block key without fetching the block
             // So the repo starts with LazyLoadingOn = true, however when not anymore in IBD 
             // a read is normally done when a peer is asking for the entire block (not just the key) 
             // then if LazyLoadingOn = false the read will be faster on the entire block            
 
-            while (!token.IsCancellationRequested)
-			{
-				if (this.StoredBlock.Height >= this.ChainState.HighestValidatedPoW?.Height)
-					break;
+            var steps = new BlockStoreLoopStepChain(disposeMode);
+            steps.SetNextStep(new ReorganiseBlockRepositoryStep(this));
+            steps.SetNextStep(new CheckNextChainedBlockExistStep(this));
+            steps.SetNextStep(new ProcessPendingStorageStep(this));
+            steps.SetNextStep(new DownloadBlockStep(this));
+
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                if (this.StoredBlock.Height >= this.ChainState.HighestValidatedPoW?.Height)
+                    break;
 
-				// find next block to download
-				var next = this.chain.GetBlock(this.StoredBlock.Height + 1);
-				if (next == null)
-					break; //no blocks to store
+                ChainedBlock nextChainedBlock = this.Chain.GetBlock(this.StoredBlock.Height + 1);
+                if (nextChainedBlock == null)
+                    break;
 
                 if (this.blockStoreStats.CanLog)
-                {
                     this.blockStoreStats.Log();
-                }                
-
-                // reorg logic
-                if (this.StoredBlock.HashBlock != next.Header.HashPrevBlock)
-				{
-					if (disposemode)
-						break;
 
-					var blockstoremove = new List<uint256>();
-					var remove = this.StoredBlock;
-					// reorg - we need to delete blocks, start walking back the chain
-					while (this.chain.GetBlock(remove.HashBlock) == null)
-					{
-						blockstoremove.Add(remove.HashBlock);
-						remove = remove.Previous;
-					}
+                BlockStoreLoopStepResult result = await steps.Execute(nextChainedBlock, cancellationToken);
```

Comment:
Are you happy with how we are executing the steps now? I like iterating over the steps approach ;)